app/src/main/java/io/gnosis/safe/ui/safe/SafeSelectionDialog.kt

```diff
@@ -0,0 +1,277 @@
+
+        viewModel.loadSafes()
+    }
+
+    companion object {
+
+        private val TAG = SafeSelectionDialog::class.java.simpleName
+
+        fun show(context: Context) {
+            val dialog = SafeSelectionDialog()
+            dialog.show((context as FragmentActivity).supportFragmentManager, TAG)
+        }
+    }
+}
+
+sealed class SafeSelectionState : BaseStateViewModel.State {
+
+    data class SafeListState(
+        val listItems: List<Any>,
+        val activeSafe: Safe?,
+        override var viewAction: BaseStateViewModel.ViewAction?
+    ) : SafeSelectionState()
+
+    data class AddSafeState(
+        override var viewAction: BaseStateViewModel.ViewAction?
+    ) : SafeSelectionState()
+}
+
+
+class SafeSelectionViewModel @Inject constructor(
+    repositories: Repositories,
+    appDispatchers: AppDispatchers
+) : BaseStateViewModel<SafeSelectionState>(appDispatchers), SafeSelectionAdapter.OnSafeSelectionItemClickedListener {
+
+    private val safeRepository = repositories.safeRepository()
+
+    private val items: MutableList<Any> = mutableListOf()
+    private var activeSafe: Safe? = null
+
+    override val state: LiveData<SafeSelectionState> = liveData {
+        for (event in stateChannel.openSubscription())
+            emit(event)
+    }
+
+    override fun initialState(): SafeSelectionState = SafeSelectionState.SafeListState(
+        listOf(AddSafeHeader()), null, null
+    )
+
+    fun loadSafes() {
+        safeLaunch {
+            val safes = safeRepository.getSafes()
+            activeSafe = safeRepository.getActiveSafe()
+
+            items.clear()
+            items.add(AddSafeHeader())
+            items.addAll(safes)
+
+            updateState { SafeSelectionState.SafeListState(items, activeSafe, null) }
+        }
+    }
+
+    fun selectSafe(safe: Safe) {
+        safeLaunch {
+            safeRepository.setActiveSafe(safe)
+            updateState { SafeSelectionState.SafeListState(items, safe, null) }
+        }
+    }
+
+    fun addSafe() {
+        safeLaunch {
+            updateState {
+                SafeSelectionState.AddSafeState(
+                    ViewAction.NavigateTo(
+                        SafeOverviewFragmentDirections.actionSafeOverviewFragmentToAddSafeNav()
+                    )
+                )
+            }
+        }
+    }
+
+    override fun onSafeClicked(safe: Safe) {
+        selectSafe(safe)
+    }
+
+    override fun onAddSafeClicked() {
+        addSafe()
+    }
+}
+
+class SafeSelectionAdapter(
+    private val clickListener: WeakReference<OnSafeSelectionItemClickedListener>
+) : RecyclerView.Adapter<BaseSafeSelectionViewHolder>() {
+
+    private val items = mutableListOf<Any>()
+
+    var activeSafe: Safe? = null
+        set(value) {
+            field = value
+            notifyAllChanged()
+        }
+
+    fun setItems(items: List<Any>, activeSafe: Safe?) {
+        this.activeSafe = activeSafe
+        this.items.clear()
+        this.items.addAll(items)
```

Comment:
This method sets active safe as a side effect. Would it make more sense to decouple that logic?